velox/functions/sparksql/aggregates/AverageAggregate.cpp

```diff
@@ -0,0 +1,263 @@
+        const TInput value = this->decodedRaw_.template valueAt<TInput>(0);
+        const auto numRows = rows.countSelected();
+        this->updateNonNullValue(group, numRows, TAccumulator(value) * numRows);
+      } else {
+        // Spark expects the result of partial avg to be non-nullable.
+        exec::Aggregate::clearNull(group);
+      }
+    } else if (this->decodedRaw_.mayHaveNulls()) {
+      rows.applyToSelected([&](vector_size_t i) {
+        if (!this->decodedRaw_.isNullAt(i)) {
+          this->updateNonNullValue(
+              group,
+              TAccumulator(this->decodedRaw_.template valueAt<TInput>(i)));
+        } else {
+          // Spark expects the result of partial avg to be non-nullable.
+          exec::Aggregate::clearNull(group);
+        }
+      });
+    } else if (
+        !exec::Aggregate::numNulls_ && this->decodedRaw_.isIdentityMapping()) {
+      const TInput* data = this->decodedRaw_.template data<TInput>();
+      TAccumulator totalSum(0);
+      rows.applyToSelected([&](vector_size_t i) { totalSum += data[i]; });
+      this->template updateNonNullValue<false>(
+          group, rows.countSelected(), totalSum);
+    } else {
+      TAccumulator totalSum(0);
+      rows.applyToSelected([&](vector_size_t i) {
+        totalSum += this->decodedRaw_.template valueAt<TInput>(i);
+      });
+      this->updateNonNullValue(group, rows.countSelected(), totalSum);
+    }
+  }
+
+ private:
+  void extractValuesImpl(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto vector = (*result)->as<FlatVector<TResult>>();
+    VELOX_CHECK(vector);
+    vector->resize(numGroups);
+    uint64_t* rawNulls = this->getRawNulls(vector);
+
+    TResult* rawValues = vector->mutableRawValues();
+    for (int32_t i = 0; i < numGroups; ++i) {
+      char* group = groups[i];
+      if (this->isNull(group)) {
+        vector->setNull(i, true);
+      } else {
+        auto* sumCount = this->accumulator(group);
+        if (sumCount->count == 0) {
+          // In Spark, if all inputs are null, count will be 0,
+          // and the result of final avg will be null.
+          vector->setNull(i, true);
+        } else {
+          this->clearNull(rawNulls, i);
+          rawValues[i] = (TResult)sumCount->sum / sumCount->count;
+        }
+      }
+    }
+  }
+};
+
+} // namespace
+
+exec::AggregateRegistrationResult registerAverage(const std::string& name) {
+  std::vector<std::shared_ptr<exec::AggregateFunctionSignature>> signatures;
+
+  for (const auto& inputType : {"smallint", "integer", "bigint", "double"}) {
+    signatures.push_back(exec::AggregateFunctionSignatureBuilder()
+                             .returnType("double")
+                             .intermediateType("row(double,bigint)")
+                             .argumentType(inputType)
+                             .build());
+  }
+  // Real input type in Presto has special case and returns REAL, not DOUBLE.
```

Comment:
This is Spark-specific file. Why is it referring to Presto? Are there any differences in signatures between Spark and Presto?

Reply:
Sorry, it's a missing point. Here is Spark's datatype [Average.scala#L66-L82](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Average.scala#L66-L82). For Real, the return type is Double in Spark. Fixed the signature.